### PR TITLE
feat: integrate offer button

### LINKED ISSUE #4542: Gallery Item Desktop - integrate offer

### PR DESCRIPTION
**Pull request by preschian**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇 \_ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring
- [x] Redesign

## Context

- [x] Closes #4542

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [ ] I've tested it at </bsx/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [ ] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=<My_Kusama_Address_check_https://github.com/kodadot/nft-gallery/blob/main/CONTRIBUTING.md#creating-your-ksm-address>)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

**Comment by roiLeo** [comment]
> I need your help @roiLeo @Jarsen136. Does the offer button work on your side on this PR? I just followed the `submit()` function on `AvailableActions`. If not, What do I miss anyway?

wfm, block hash [0xa537886e39455a2d6e0cde1f96177433b5d4c65d996422f9abf0dc1431c6c9ad](https://polkadot.js.org/apps/#/explorer/query/0xa537886e39455a2d6e0cde1f96177433b5d4c65d996422f9abf0dc1431c6c9ad)  

it's been a few weeks that Snek [Rococo] endpoint is "stuck" but you can see that the call has been made on chain

**Comment by Jarsen136** [comment]
> I need your help @roiLeo @Jarsen136. Does the offer button work on your side on this PR? I just followed the `submit()` function on `AvailableActions`. If not, What do I miss anyway?

WFM too

File: composables/useToast.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// vue composables api for toast message function
+export default function useToast() {
```
Comment (Jarsen136) [inline composables/useToast.ts:2]: It seems that this composables method was not being used anymore. Should we remove it ?

**Comment by yangwao** [comment]
> Snek [Rococo] endpoint is "stuck"

After Brick I sense @vikiival updated only snek@05 for BSX, not for Rococo?

**Comment by vikiival** [comment]
Looks like something is wrong with offers. I pinged BSX team

**Comment by yangwao** [comment]
I've made offer on this one https://deploy-preview-4554--koda-nuxt.netlify.app/bsx/gallery/2608295324-7

But it seems on old one doesn't show up. 
image

https://basilisk.subscan.io/extrinsic/2460385-2

image

Oh, seems it's on Basilisk team side something ...

**Comment by vikiival** [comment]
Offers should be already available

**Comment by roiLeo** [comment]
> Offers should be already available

Tx is good but it's not visible for me on snek Rococo

**Comment by vikiival** [comment]
> Tx is good but it's not visible for me on snek Rococo

Because the one that @yangwao made was on BSX

File: composables/useTransaction.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  return {
+    isLoading,
+    status,
+    executeTransaction,
+  }
+}
+
+export const useTransaction = () => {
```
[the remark this answers is not in the document]
Comment (roiLeo) [reply composables/useTransaction.ts:34]:
```ts
  const transactionMap = (item: Actions, api: ApiPromise) => {
    return {
      [Interaction.LIST]: execTsxList(item as ActionList, api, executeTransaction),
      [Interaction.SEND]: execTsxSend(item as ActionSend, api, executeTransaction),
      [ShoppingActions.MAKE_OFFER]: execTsxOffer(item as ActionOffer, api, executeTransaction),
    }[item.interaction] || 'UNKNOW'
  }

  const transaction = async (item: Actions) => {
    const api = await apiInstance.value

    return transactionMap(item, api)
  }
```
no ts error lmk what you think

[the remark this answers is not in the document]
Comment (roiLeo) [reply composables/useTransaction.ts:34]: nevermind I'll try to debug this

**Review by roiLeo — COMMENTED** [review]
On dark mode text is hard to see
Screenshot 2023-01-12 at 10-29-50 KodaDot - Kusama NFT Market Explorer

I've tested it on `/snek/gallery/659233203-7?redesign=true` and when I log `item` I find a wrong tokenId

```json
{
  "interaction": "MAKE_OFFER",
  "currentOwner": "bXhWeVLtArWauM8UmbaVdDmXmo5uSDk6DAa2YxQ5zbt2sDcbW",
  "day": 3,
  "price": ".1",
  "tokenId": "659233203-659233203-7",
  "urlPrefix": "snek",
  "successMessage": "Offer successfully created",
  "errorMessage": "Failed to send item"
}
```

File: composables/useTransaction.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (roiLeo) [reply composables/useTransaction.ts:34]:
```ts
  const transactionMap = (item: Actions, api: ApiPromise) => {
    const map = {
      [Interaction.LIST]: (item, api) => execTsxList(item as ActionList, api, executeTransaction),
      [Interaction.SEND]: (item, api) => execTsxSend(item as ActionSend, api, executeTransaction),
      [ShoppingActions.MAKE_OFFER]: (item, api) => execTsxOffer(item as ActionOffer, api, executeTransaction)
    }

    return map[item.interaction]?.(item, api) ?? 'UNKNOW'
  }
```

or shorthand

```ts
  const transactionMap = (item: Actions, api: ApiPromise) =>
    ({
      [Interaction.LIST]: (item, api) =>
        execTsxList(item as ActionList, api, executeTransaction),
      [Interaction.SEND]: (item, api) =>
        execTsxSend(item as ActionSend, api, executeTransaction),
      [ShoppingActions.MAKE_OFFER]: (item, api) =>
        execTsxOffer(item as ActionOffer, api, executeTransaction),
    }[item.interaction]?.(item, api) ?? 'UNKNOW')
```

**Comment by roiLeo** [comment]
✅ Block hash: 0xe77afe69d0712c73bfc45df5a9ac5300ba46d22bb05bf548bfd281ab24e76546 on Rococo

Still blocked by kodadot/snek#130

**Review by Jarsen136 — APPROVED** [review]
code lgtm

**Comment by vikiival** [comment]
Merged

https://github.com/kodadot/snek/pull/136

**Comment by roiLeo** [comment]
> Merged
> 
> [kodadot/snek#136](https://github.com/kodadot/snek/pull/136)

https://squiki.kodadot.workers.dev/status/snekk

new endpoint again??? (`https://squid.subsquid.io/snekk/graphql`)

**Comment by vikiival** [comment]
> new endpoint again???

Following good practises 
https://docs.subsquid.io/deploy-squid/promote-to-production/

**Comment by roiLeo** [comment]
Yeah!! offers are back
Screenshot 2023-01-12 at 16-06-34 Golden Gate KodaDot - Kusama NFT Market Explorer

**Review by roiLeo — APPROVED** [review]
✅ wfm 🚀

File: composables/transaction/transactionOffer.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { getExpiration } from '@/utils/api/bsx/query'
+import { calculateBalance } from '@/utils/format/balance'
+import { bsxParamResolver, getApiCall } from '@/utils/gallery/abstractCalls'
+import { dangerMessage } from '@/utils/notification'
+import { ShoppingActions } from '@/utils/shoppingActions'
+
+import type { ActionOffer } from './types'
+
+export async function execTsxOffer(item: ActionOffer, api, executeTransaction) {
```
Comment (vikiival) [inline composables/transaction/transactionOffer.ts:9]: execTsxOffer

`exec{What}Tx` ?

```suggestion
export async function exectMakeOfferTx(item: ActionOffer, api, executeTransaction) {
```

**Review by vikiival — CHANGES_REQUESTED** [review]
Smol

**Review by vikiival — APPROVED** [review]
let it roll 🎲

**Pull request merged**